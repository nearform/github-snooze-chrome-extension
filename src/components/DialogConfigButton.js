import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Typography,
  TextField,
  Box
} from '@mui/material'
import SettingsIcon from '@mui/icons-material/Settings'
import React, { useState } from 'react'
import T from 'prop-types'
import { createChromeAlarm } from '../background'
import { useChromeLocalStorage } from '../hooks/useChromeLocalStorage'

const SETTINGS_LABELS = {
  checkIntervalTimer: 'Check Interval Timer (minutes)'
}

const rowStyles = {
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  justifyContent: 'space-around'
}

const ConfigurationTitle = ({ title }) => {
  return (
    <Box sx={{ maxWidth: '50%' }}>
      <Typography fontSize={14}>{title}</Typography>
    </Box>
  )
}

const ConfigurationField = ({ renderField }) => {
  return <Box sx={{ maxWidth: '20%' }}>{renderField()}</Box>
}

export default function DialogConfigButton({ title, disabled }) {
  const [open, setOpen] = useState(false)
  const {
    setData: setCheckIntervalTimer,
    localData: localCheckIntervalTimer,
    setLocalData: setLocalCheckIntervalTimer
  } = useChromeLocalStorage('checkIntervalTimer')

  const handleClickOpen = () => setOpen(true)
  const handleDialogClose = () => setOpen(false)

  if (!localCheckIntervalTimer) {
    return
  }

  const handleCheckIntervalTimerValidation = event => {
    const { value } = event.target
    setLocalCheckIntervalTimer(value)
  }

  const handleApplyChanges = () => {
    setCheckIntervalTimer(localCheckIntervalTimer)
    createChromeAlarm(Number(localCheckIntervalTimer))
    handleDialogClose()
  }

  return (
    <div>
      <Button
        color="secondary"
        variant="outlined"
        size="small"
        data-testid={'open-settings-dialog'}
        onClick={handleClickOpen}
        disabled={disabled}
      >
        <SettingsIcon />
      </Button>
      <Dialog open={open} onClose={handleDialogClose}>
        <DialogTitle>{title}</DialogTitle>
        <DialogContent>
          <Box sx={rowStyles}>
            <ConfigurationTitle title={SETTINGS_LABELS.checkIntervalTimer} />
            <ConfigurationField
              renderField={() => (
                <TextField
                  type="number"
                  inputProps={{
                    min: 1
                  }}
                  value={localCheckIntervalTimer}
                  onChange={handleCheckIntervalTimerValidation}
                  size="small"
                  variant="outlined"
                />
              )}
            />
          </Box>
        </DialogContent>
        <DialogActions>
          <Button color="secondary" size="small" onClick={handleDialogClose}>
            Close
          </Button>
          <Button
            variant="contained"
            color="secondary"
            size="small"
            onClick={handleApplyChanges}
          >
            Apply
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  )
}

DialogConfigButton.propTypes = {
  title: T.string.isRequired,
  size: T.string.isRequired,
  disabled: T.bool.isRequired
}
