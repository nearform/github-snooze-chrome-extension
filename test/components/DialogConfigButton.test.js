import React from 'react'
import { fireEvent } from '@testing-library/react'
import { render } from '../renderer'
import DialogConfigButton from '../../src/components/DialogConfigButton'

const props = {
  title: 'title',
  disabled: false
}

const DATA_TEST_ID_OPEN_SETINGS_DIALOG = 'open-settings-dialog'
const DATA_TEST_ID_CHECK_INTERVAL_INPUT = 'input-check-interval-timer'

const mockSetData = jest.fn()
const mockSetLocalData = jest.fn()

jest.mock('../../src/hooks/useChromeLocalStorage.js', () => ({
  useChromeLocalStorage: () => ({
    data: 1,
    setData: mockSetData,
    localData: 1,
    setLocalData: mockSetLocalData
  })
}))

describe('DialogConfigButton.js', () => {
  test('shows the proper enabled DialogFormButton', async () => {
    const { asFragment, findByTestId } = render(
      <DialogConfigButton {...props} />
    )
    await findByTestId(DATA_TEST_ID_OPEN_SETINGS_DIALOG)
    expect(asFragment()).toMatchSnapshot()
  })

  test('shows the proper disabled DialogFormButton', async () => {
    const { asFragment, findByTestId } = render(
      <DialogConfigButton {...props} disabled={true} />
    )

    await findByTestId(DATA_TEST_ID_OPEN_SETINGS_DIALOG)

    expect(asFragment()).toMatchSnapshot()
  })

  test('should apply change in settings - check interval timer', async () => {
    const { findByTestId, findByText } = render(
      <DialogConfigButton {...props} />
    )

    const settingsButton = await findByTestId(DATA_TEST_ID_OPEN_SETINGS_DIALOG)
    fireEvent.click(settingsButton)

    const checkIntervalTimerInput = await findByTestId(
      DATA_TEST_ID_CHECK_INTERVAL_INPUT
    )
    const newTimerValue = +checkIntervalTimerInput.value + 1

    fireEvent.change(checkIntervalTimerInput, {
      target: { value: newTimerValue }
    })
    expect(mockSetLocalData).toHaveBeenCalledTimes(1)

    const applyButton = await findByText('Apply')
    fireEvent.click(applyButton)

    expect(mockSetData).toHaveBeenCalledTimes(1)
  })
})
