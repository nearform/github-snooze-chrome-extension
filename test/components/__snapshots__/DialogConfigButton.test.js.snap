// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogConfigButton.js shows the proper disabled DialogFormButton 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeSmall MuiButton-outlinedSizeSmall Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-nnjl0l-MuiButtonBase-root-MuiButton-root"
      data-testid="open-settings-dialog"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SettingsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`DialogConfigButton.js shows the proper enabled DialogFormButton 1`] = `
<DocumentFragment>
  <div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeSmall MuiButton-outlinedSizeSmall css-nnjl0l-MuiButtonBase-root-MuiButton-root"
      data-testid="open-settings-dialog"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="SettingsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
